[5.4] Installation - Admin password must be at least 12 characters long

alikon opened the pull request:
Pull Request for Issue #46111, #42289 .

### Summary of Changes

Admin password must be at least 12 characters long and Password must not have spaces at the beginning or end

### Testing Instructions

Install Joomla 5.3.x
At Login Data: set a short <12 char  and or use spaces as 'Set the username for your Super User account. *'


### Actual result BEFORE applying this Pull Request

The installation process goes on until it gets stuck in the 'Installation progress' phase. There is no possibility to cancel, stop or do something other than go back in the browser.

### Expected result AFTER applying this Pull Request

 you cannot proceed untill you set a minimum password of 12 chars and Password must not have spaces at the beginning or end
<img width="1324" height="1292" alt="image" src="https://github.com/user-attachments/assets/0b444608-5533-498b-95bd-48ce478b5dd8" />



### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

alikon committed:
Admin password must be at least 12 characters long

tecpromotion commented:
@alikon Please do not use hard-coded strings. This should be a variable so that appropriate translations can be made in all languages.

alikon committed:
translatable error message

alikon committed:
translatable error message

brianteeman commented:
I have tested this item :white_check_mark: successfully on 65a2286ae6b33c972d3ff6345f7de0d218bec6aa<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/46173">issues.joomla.org/tracker/joomla-cms/46173</a>.</sub>

alikon committed:
check spaces

alikon committed:
check spaces

tecpromotion commented:
It works in principle, but the string `JFIELD_PASSWORD_SPACES_IN_PASSWORD` is not entirely correct, as spaces within the password are not accepted either. However, it only refers to the beginning and end.

alikon commented:
then the text of `JFIELD_PASSWORD_SPACES_IN_PASSWORD="Password must not have spaces at the beginning or end."`  is wrong

as it is only used here https://github.com/joomla/joomla-cms/blob/1b14ada694a179a144ad53f8992ab8739a97e7c9/libraries/src/Form/Rule/PasswordRule.php#L95-L115

richard67 commented on installation/template/js/setup.js:
```diff
@@ -164,6 +179,24 @@ Joomla.checkDbCredentials = function() {
+      }
+    });
+  }
+
```
Hmm, to of these blocks `// Clear error messages when password becomes valid` for almost the same thing, first one lines 182 to 189, then 2nd one lines 191 to 199? Is that by purpose or is it a mistake?

alikon replied:
better error management

alikon committed:
shoe exact errors

HLeithner commented:
This pull request has been automatically rebased to 5.4-dev.

wojsmol commented:
@richard67 prebuild packages for this PR aren't available.

richard67 committed:
Merge branch '5.4-dev' into patch-32

richard67 commented:
> @richard67 prebuild packages for this PR aren't available.

Seems they were too old and so deleted. I have just triggered a new build, they should be available again in some 30 minutes.

richard67 committed:
Merge branch '5.4-dev' into patch-32

alikon committed:
Merge branch '5.4-dev' into patch-32

muhme committed:
Merge branch '5.4-dev' into patch-32

muhme commented:
Tested with JBT and en-GB
* Before the PR
  * There is a note 'Enter at least 12 characters.' and a orange hint 'Password doesn't meet the site's requirements.' But if I use a shorter password, I can continue. Installation fails later with clear red error message "Password is too short." – But of course, it would be better to fail directly on the password input page.
  * With one space in the middle the installation is working.
  * With only 12 spaces as password the installation fails later with clear red error message 'Password must not have spaces at the beginning or end.'  – But again, it would be better to fail directly on the password input page.
* Graft PRs full package and deleted configuration.php
  * Tested password shorter than 12 chars have red error 'Enter at least 12 characters.'
    * ⚠️ But with typing each character the error message pops-up new as nervous fidgeting. Is it possible to check more user-friendly on 'Setup Database Connection' click? See ![demo](https://github.com/user-attachments/assets/26a39c40-ad9c-4346-9ade-31e3faf023ed)
    * ❌ Test fails as even with a space in the middle, which was working before, there is the error message 'Password must not have spaces at the beginning or end.'

alikon committed:
Merge branch '5.4-dev' into patch-32

alikon committed:
onblur

alikon commented:
> but with typing each character the error message pops-up new as nervous fidgeting. Is it possible to check more user-friendly

used on blur now

> Test fails as even with a space in the middle, which was working before, there is the error message 'Password must not have spaces at the beginning or end.'

see previous comment  _"We don't allow white space inside passwords"_ 
https://github.com/joomla/joomla-cms/pull/46173#issuecomment-3338404880

muhme commented:
@alikon Thank you very much for further improving this PR. We discussed this PR in yesterday's maintainer meeting. Since white spaces within passwords currently work  (we have tested creating a user, login and changing the password), we would not change this functionality in a minor release. Could you please adjust the PR to that it only checks for spaces in the beginning and end of a password?

tecpromotion commented:
> @alikon Thank you very much for further improving this PR. We discussed this PR in yesterday's maintainer meeting. Since white spaces within passwords currently work (we have tested creating a user, login and changing the password), we would not change this functionality in a minor release. Could you please adjust the PR to that it only checks for spaces in the beginning and end of a password?

Spaces within passwords are expected behaviour; there are password managers that generate such passwords, and this should continue to work.

alikon committed:
Merge branch '5.4-dev' into patch-32

alikon committed:
Changed the password validation to only block spaces at the beginning and end of the password

alikon committed:
Merge branch '5.4-dev' into patch-32

alikon committed:
Merge branch '5.4-dev' into patch-32

laoneo commented:
https://github.com/joomla/joomla-cms/blob/5.4-dev/build/media_source/system/js/fields/passwordstrength.es6.js

alikon committed:
undo

alikon committed:
check for 12 chars

alikon committed:
unneeded

laoneo commented on build/media_source/system/js/fields/passwordstrength.es6.js:
```diff
@@ -157,32 +135,46 @@ class PasswordStrength {
+        const trimmedValue = value.trim();
+        
+        // Block if less than 12 characters
+        if (trimmedValue.length < 12) {
```
I think here should the minLength parameter be used. Otherwise you wont allow user passwords < 12, even when allowed in Joomla.

alikon replied:
now it care of minLength

alikon committed:
take count of minLength

alikon committed:
Merge branch '5.4-dev' into patch-32

laoneo commented:
Just wondering, did the CS changes a tool or is it your preference?

alikon commented:
both VScode + preference

alikon committed:
better code

alikon committed:
Implement password strength validation tests

Added tests for password strength validation and handling of optional fields during installation.

alikon commented:
added some system test

laoneo commented on tests/System/integration/install/Installation.cy.js:
```diff
@@ -1,3 +1,50 @@
+
+  it('allows empty db_pass (optional field)', () => {
+    cy.get('#step1').click();
+    cy.get('#jform_admin_user').clear().type('admin');
```
Please use here Cypress.env('username')

alikon replied:
done

laoneo commented on tests/System/integration/install/Installation.cy.js:
```diff
@@ -1,3 +1,50 @@
+    cy.get('#jform_admin_user').clear().type('admin');
+    cy.get('#jform_admin_username').clear().type('ci-admin');
+    cy.get('#jform_admin_password').clear().type('ValidPass123!');
+    cy.get('#jform_admin_email').clear().type('admin@example.com');
```
Please use Cypress.env('email')

alikon replied:
done

alikon committed:
Merge branch '5.4-dev' into patch-32

alikon committed:
Replace hardcoded values with Cypress environment variables

alikon committed:
Merge branch '5.4-dev' into patch-32

alikon committed:
Merge branch '5.4-dev' into patch-32

alikon commented:
it will need more care and more time
thank you all for the feedback

alikon committed:
Merge branch '5.4-dev' into patch-32

alikon committed:
Merge branch '5.4-dev' into patch-32

alikon committed:
fix typo in password strength field and improve setup script

alikon committed:
Merge branch '5.4-dev' into patch-32

alikon committed:
split

alikon committed:
original Installation.cy.js

alikon committed:
add cleaning

alikon committed:
test

alikon committed:
clean db

alikon committed:
clean first

alikon committed:
clear

alikon committed:
order

alikon committed:
session clean

alikon committed:
small delay

alikon committed:
force page reload

tecpromotion commented:
@alikon it would be great if we could get that in 5.4.4 so that it's also included in 6.1.0. Thank you for staying on top of this issue.

alikon committed:
fix tests

alikon committed:
no change

alikon committed:
disable tests for now

alikon committed:
simplify

alikon committed:
re-tests

alikon committed:
Merge branch '5.4-dev' into patch-32

alikon committed:
Merge branch '5.4-dev' into patch-32

alikon committed:
Merge branch '5.4-dev' into patch-32

alikon commented:
@tecpromotion ready for human test

alikon committed:
message when trailing spaces

richard67 commented on build/media_source/system/js/fields/passwordstrength.es6.js:
```diff
@@ -45,27 +45,26 @@ class PasswordStrength {
       }
     });
-
+  
```
```suggestion

```
Please remove spaces in empty lines.

richard67 commented on build/media_source/system/js/fields/passwordstrength.es6.js:
```diff
@@ -45,27 +45,26 @@ class PasswordStrength {
+      mods
     );
-
+  
```
```suggestion

```
Please remove spaces in empty lines.

alikon committed:
Update build/media_source/system/js/fields/passwordstrength.es6.js

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

alikon committed:
Update build/media_source/system/js/fields/passwordstrength.es6.js

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

alikon committed:
Add message for spaces in password validation

alikon committed:
Update password strength warning message

alikon committed:
Merge branch '5.4-dev' into patch-32
